pex/resolver.py

```diff
@@ -248,8 +248,12 @@ def package_iterator(self, resolvable, existing=None):
 
       if self.__cache_ttl:
         packages = self.filter_packages_by_ttl(packages, self.__cache_ttl)
-        if packages:
-          return packages
+        # Return both cached packages and the pypi packages. The edge case is if a inexact
+        # resolvable finds a cached version, it needs to return that cached package AND all pypi
+        # packages because the cached version might not match another resolvable which has a tighter
+        # bound or an exact version.
+        return packages + super(CachingResolver, self).package_iterator(resolvable,
+                                                                        existing=existing)
 
     return super(CachingResolver, self).package_iterator(resolvable, existing=existing)
```

Comment:
Gotcha, yah that looks good to me. I'll push up the change. For now I'll do a separate autosquashable commit unless you'd prefer it be a separate commit. Local tests run fine with that change btw :)